lib/devfs/device.c

```diff
@@ -77,95 +78,75 @@ device_lookup(const char *name)
 	return NULL;
 }
 
-struct partition_table_entry {
-	uint8_t  bootable;
-	uint8_t  starting_head;
-	uint16_t starting_sector:6;
-	uint16_t starting_cylinder:10;
-	uint8_t  system_id;
-	uint8_t  ending_head;
-	uint16_t ending_sector:6;
-	uint16_t ending_cylinder:10;
-	uint32_t rela_sector;
-	uint32_t total_sectors;
-} __packed;
-
-
-void device_register(struct device *dev, const char *name, int flags)
-{
-	size_t len;
-	void *priv = NULL;
-
-	UK_ASSERT(dev->driver != NULL);
-
-	/* Check the length of name. */
-	len = strnlen(name, MAXDEVNAME);
-	if (len == 0 || len >= MAXDEVNAME)
-		return;
-
-	uk_mutex_lock(&devfs_lock);
-
-	/* Check if specified name is already used. */
-	if (device_lookup(name) != NULL)
-		UK_CRASH("duplicate device");
-
-	/*
-	 * Allocate a device and device private data.
-	 */
-	if (dev->driver->devsz != 0) {
-		if ((priv = malloc(dev->driver->devsz)) == NULL)
-			UK_CRASH("devsz");
-		memset(priv, 0, dev->driver->devsz);
-	}
-
-	strlcpy(dev->name, name, len + 1);
-	dev->flags = flags;
-	dev->active = 1;
-	dev->refcnt = 1;
-	dev->offset = 0;
-	dev->private_data = priv;
-	dev->next = device_list;
-	dev->max_io_size = UINT_MAX;
-	device_list = dev;
-
-	uk_mutex_unlock(&devfs_lock);
-}
-
-
 /*
  * device_create - create new device object.
  *
  * A device object is created by the device driver to provide
  * I/O services to applications.  Returns device ID on
  * success, or 0 on failure.
```

Comment:
This is wrong. It doesn't return `0` on error. It returns a positive (?) error code.

Also, given the `checkpatch` complaint: why is the return value positive (i.e. `ENOMEM` instead of `-ENOMEM`)?

Reply:
Hey @razvand,

good point. I fixed the stale comment. It was already broken before. Unfortunately, all functions in devfs happen to return positive errno values (🤮) because they are used in `devfs_vnops.c` to handle vnode operations. At this point, vfscore still expects positive error codes so I stayed with the positive codes here, too.

The fdtab #683 PR changes this so that vfscore finally expects negative errno values to denote errors. The PR changes this also for `device_create` and the other functions in `devfs`. So the PR just needs to be rebased.